src/v/storage/disk_log_impl.cc

```diff
@@ -1731,6 +1733,446 @@ disk_log_impl::make_cached_reader(log_reader_config config) {
+            vlog(
+              stlog.debug,
+              "Adding {} bytes to the offset range. Segment offsets: {}, "
+              "current_size: {}",
+              sz,
+              it->get()->offsets(),
+              current_size);
+        }
+        segments.push_back(*it);
+        if (current_size > target.target_size) {
+            // Segment size overshoots and we need to use segment index
+            // to find the end offset and its file pos.
+            // We accumulated enough segments to satisfy the query. There're
+            // few cases:
+            // - The beginning and the end of the range is inside the same
+            // segment. The 'segments' collection will have only one element
+            // if this is the case.
+            // - The beginning and the end are located in different
+            // segments. In both cases we need to find the finish line. The
+            // size calculation is affected by these two possibilities.
+            vlog(
+              stlog.debug,
+              "Offset range size overshoot by {}, current segment size "
+              "{}, current offset range size: {}",
+              current_size - target.target_size,
+              it->get()->file_size(),
+              current_size);
+
+            size_t truncate_after = 0;
+            if (segments.size() == 1) {
+                // At this point it's guaranteed that 'base_index_entry'
+                // contains value and the 'current_size' contains 'file_size
+                // - base_index_entry->file_pos'. The expression below is
+                // guaranteed to not underflow. The 'file_size - filepos' is
+                // greater than 'range.max_size' because 'current_size >
+                // 'range.max_size'.
+                truncate_after = it->get()->file_size() - base_file_pos
+                                 - target.target_size;
+            } else {
+                auto prev = current_size - it->get()->file_size();
+                auto delta = current_size - target.target_size;
+                truncate_after = prev + delta;
+            }
+            auto last_index_entry = it->get()->index().find_above_size_bytes(
+              truncate_after);
+            if (
+              last_index_entry.has_value()
+              && model::prev_offset(last_index_entry->offset) > first) {
+                vlog(
+                  stlog.debug,
+                  "Setting offset range to {} - {}, {} bytes of the last "
+                  "segments are included",
+                  first,
+                  last_index_entry->offset,
+                  truncate_after);
+
+                last_included_offset = model::prev_offset(
+                  last_index_entry->offset);
+
+                // We're including only part of the file so the total size
+                // of the range has to be adjusted
+                current_size
+                  -= (it->get()->file_size() - last_index_entry->filepos);
```

Comment:
Maybe this is true with the current implementation, but doesn't this yield slightly inaccurate results? If the uploaded range of offsets isn't aligned to the granularity of the index, doesn't this fudge the resulting size a bit in a way that results in sizes that are slightly off?

Reply:
Here we're stopping at the index boundary. The idea is that when we're doing normal uploads we have a last uploaded offset and the target size of the segment to upload (which is usually in megabytes). We will find the first offset of the upload precisely by using index and then scanning. But in order to find the last offset we don't need such precision. We can just use index to find some offset that justifies our requirements. If it's off by few hundred bytes it's not a problem. 

The good thing it that we don't need to scan the segment. This has the added benefit. After the first segment is uploaded all subsequent uploads start on an index boundary which means that we don't need to scan segments in the future. All subsequent uploads will begin and end on index boundaries.